driver-core/src/main/com/mongodb/internal/connection/ConcurrentPool.java

```diff
@@ -197,28 +197,27 @@ public void prune() {
      * The {@code postCreate} action throwing a exception causes this method to stop and re-throw that exception.
      *
      * @param initialize An action applied to non-{@code null} new items.
-     *                   If this action throws an {@link Exception}, it must release resources associated with the item.
+     *                   If an exception is thrown by the action, the action must treat the provided item as if obtained via
+     *                   a {@link #get(long, TimeUnit) get…} method, {@linkplain #release(Object, boolean) releasing} it
+     *                   if an exception is thrown; otherwise the action must not release the item.
      */
     public void ensureMinSize(final int minSize, final Consumer<T> initialize) {
         while (getCount() < minSize) {
             if (!acquirePermit(0, TimeUnit.MILLISECONDS)) {
                 break;
             }
-            T newItem = createNewAndReleasePermitIfFailure(initialize);
+            T newItem = createNewAndReleasePermitIfFailure();
+            initialize.accept(newItem);
             release(newItem);
```

Comment:
Should this be in a finally block?

Reply:
No, it's important for this method to not always be called:

1. if creation failed, `createNewAndReleasePermitIfFailure` releases the permit;
2. if initialization failed, the initializer (`initialize`) releases the item as per the contract of the `ensureMinSize` method; (this contract seems to be the lesser of all evils, allowing to not further complicate the implementation of `initialize`)
3. `release(newItem)` is called only when both creation of `newItem` and initialization succeeded;

We still can put `release(newItem)` in a `finally` block like this:
```java
T newItem = null;
boolean success = false;
if (!acquirePermit(0, TimeUnit.MILLISECONDS)) {
    break;
}
try {
    newItem = createNewAndReleasePermitIfFailure();
    initialize.accept(newItem);
    success = true;
} finally {
    if (success) {
        release(newItem);
    }
}
```

Reply:
Thanks for the explanation - no need for any code change.